tuplex/codegen/src/IteratorContextProxy.cc

```diff
@@ -0,0 +1,494 @@
+            }
+
+            auto* dl = new DataLayout(_env->getModule().get());
+            return SerializableValue(iteratorContextStruct, _env->i64Const(dl->getTypeAllocSize(iteratorContextType)));
+        }
+
+        SerializableValue IteratorContextProxy::initZipContext(LambdaFunctionBuilder &lfb, llvm::IRBuilder<> &builder,
+                                                           const std::vector<SerializableValue> &iterables,
+                                                           IteratorInfo *iteratorInfo) {
+            using namespace llvm;
+
+            if(iterables.empty()) {
+                // use dummy value for empty iterator
+                return SerializableValue(_env->i64Const(0), _env->i64Const(8));
+            }
+
+            auto iterablesType = iteratorInfo->argsType;
+            auto argsIteratorInfo = iteratorInfo->argsIteratorInfo;
+            llvm::Type *iteratorContextType = _env->getZipIteratorType(iterablesType, argsIteratorInfo);
+            if(iteratorContextType == _env->i64Type()) {
+                // empty iterator
+                return SerializableValue(_env->i64Const(0), _env->i64Const(8));
+            }
+            auto iteratorContextStruct = _env->CreateFirstBlockAlloca(builder, iteratorContextType, "zip_iterator_alloc");
+            // store pointers to iterator structs
+            for (int i = 0; i < iterablesType.parameters().size(); ++i) {
+                auto currType = iterablesType.parameters()[i];
+                assert(currType.isIterableType());
+                auto iterablePtr = builder.CreateGEP(iteratorContextType, iteratorContextStruct, {_env->i32Const(0), _env->i32Const(i)});
+                llvm::Value *iteratorVal;
+                if(currType.isIteratorType()) {
+                    iteratorVal = iterables[i].val;
+                } else {
+                    iteratorVal = initIterContext(lfb, builder, currType, iterables[i]).val;
+                }
+                builder.CreateStore(iteratorVal, iterablePtr);
+            }
+
+            auto* dl = new DataLayout(_env->getModule().get());
+            return SerializableValue(iteratorContextStruct, _env->i64Const(dl->getTypeAllocSize(iteratorContextType)));
+        }
+
+        SerializableValue IteratorContextProxy::initEnumerateContext(LambdaFunctionBuilder &lfb,
+                                                                     llvm::IRBuilder<> &builder,
+                                                                     const SerializableValue &iterable,
+                                                                     llvm::Value *startVal,
+                                                                     IteratorInfo *iteratorInfo) {
+            using namespace llvm;
+
+            auto iterableType = iteratorInfo->argsType;
+            if(iterableType == python::Type::EMPTYITERATOR || iterableType == python::Type::EMPTYLIST || iterableType == python::Type::EMPTYTUPLE) {
```

Comment:
empty dict?

Reply:
Dictionary is handled through fallback.
Added check and error message.